clippy_lints/src/methods/iter_out_of_bounds.rs

```diff
@@ -0,0 +1,91 @@
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::{is_trait_method, match_def_path, paths};
+use rustc_ast::LitKind;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{self};
+use rustc_span::sym;
+
+use super::ITER_OUT_OF_BOUNDS;
+
+/// Attempts to extract the length out of an iterator expression.
+fn get_iterator_length<'tcx>(cx: &LateContext<'tcx>, iter: &'tcx Expr<'tcx>) -> Option<u128> {
+    let iter_ty = cx.typeck_results().expr_ty(iter);
+
+    if let ty::Adt(adt, substs) = iter_ty.kind() {
```

Comment:
nit: This could probably be `let-else`.

Reply:
nit: I think we can also call `ty_adt_def` instead.

`peel_refs` is also probably ok too, but I don't think there are many cases where that'll even come up if it changes anything (why would somebody do `(&[1, 2, 3].iter()).skip(x)`?)

Reply:
> nit: I think we can also call `ty_adt_def` instead.

I don't think so, since we need the generic args, and we don't get those with `ty_adt_def`

Reply:
Ah, missed that that was used. That's fine then